csharp/source/autowrap/csharp/csharp.d

```diff
@@ -0,0 +1,1061 @@
+    else if (type == byteTypeString) return "sbyte";
+    else if (type == ushortTypeString) return "ushort";
+    else if (type == shortTypeString) return "short";
+    else if (type == uintTypeString) return "uint";
+    else if (type == intTypeString) return "int";
+    else if (type == ulongTypeString) return "ulong";
+    else if (type == longTypeString) return "long";
+    else if (type == floatTypeString) return "float";
+    else if (type == doubleTypeString) return "double";
+    else return getCSharpName(type);
+}
+
+private string getCSharpInterfaceType(string type) {
+    //Types that require special marshalling types
+    if (type[$-2..$] == "[]") type = type[0..$-2];
+    if(type == voidTypeString) return "void";
+    else if(type == stringTypeString) return "string";
+    else if (type == wstringTypeString) return "string";
+    else if (type == dstringTypeString) return "string";
+    else if (type == boolTypeString) return "bool";
+    //else if (type == dateTimeTypeString) return "DateTime";
+    //else if (type == sysTimeTypeString) return "DateTimeOffset";
+    //else if (type == uuidTypeString) return "Guid" 
+
+    //Types that can be marshalled by default
+    else if (type == charTypeString) return "byte";
+    else if (type == wcharTypeString) return "char";
+    else if (type == dcharTypeString) return "uint";
+    else if (type == ubyteTypeString) return "byte";
+    else if (type == byteTypeString) return "sbyte";
+    else if (type == ushortTypeString) return "ushort";
+    else if (type == shortTypeString) return "short";
+    else if (type == uintTypeString) return "uint";
+    else if (type == intTypeString) return "int";
+    else if (type == ulongTypeString) return "ulong";
+    else if (type == longTypeString) return "long";
+    else if (type == floatTypeString) return "float";
+    else if (type == doubleTypeString) return "double";
+    else return getCSharpName(type);
+}
+
+private string getDLangReturnType(string type, bool isNothrow, bool isClass) {
+    import std.stdio;
+    if(isNothrow) {
+        return getDLangInterfaceType(type);
+    } else {
+        string rtname = getReturnErrorTypeName(getDLangInterfaceType(type));
+        //These types have predefined C# types.
+        if(type == voidTypeString || type == boolTypeString || type == stringTypeString || type == wstringTypeString || type == dstringTypeString || rtname == "return_slice_error") {
+            return rtname;
+        }
+
+        string typeStr = "    [StructLayout(LayoutKind.Sequential)]" ~ newline;
+        typeStr ~= "    internal struct %s {".format(rtname) ~ newline;
+        typeStr ~= "        private void EnsureValid() {" ~ newline;
+        typeStr ~= "            var errStr = SharedFunctions.SliceToString(_error, DStringType._wstring);" ~ newline;
+        typeStr ~= "            if (!string.IsNullOrEmpty(errStr)) throw new DLangException(errStr);" ~ newline;
+        typeStr ~= "        }" ~ newline;
```

Comment:
Why this style everywhere? Why not
```D
        string typeStr =
`    [StructLayout(LayoutKind.Sequential)]
    internal struct ` ~ rtname ~ ` {
        private void EnsureValid() {
            var errStr = SharedFunctions.SliceToString(_error, DStringType._wstring);
            if (!string.IsNullOrEmpty(errStr)) throw new DLangException(errStr);
        }
`;
```
or similar?

Reply:
To be honest, because it preserves the indenting and makes it easier to see what's being generated.